### PR TITLE
PR3: Direct funding test using chain and two wallets

### PR DESCRIPTION
**Pull request by lalexgap**
**based on top of #3564 and #3555** 

Adds a very basic test that ensures direct funding works between two wallets using a chain service running against a real ganache.

Currently, the test creates and then closes a channel. It ensures that the wallet promises resolve successfully and that the funds on ganache get updated correctly.

### CODE REVIEW

File: packages/server-wallet/jest/chain-setup.ts
Comment (lalexgap) [inline packages/server-wallet/jest/chain-setup.ts:10]: This allows chain tests to write out logs to the artifact folder for circleCI

File: packages/server-wallet/jest/chain-setup.ts
Comment (kerzhner) [inline packages/server-wallet/jest/chain-setup.ts:13]: Does this prevent multiple, local runs of the chain tests?

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:134]: Should we use a logger instead of `console`?

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:188]: Does this work consistently? If so, I am a bit surprised!

I could be misreading the code, but it seems to me that we are only waiting for the objective approval, not objective succeeding.

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:195]: Do we need to wait for the objective to succeed?

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:201]: It might be worth it to set `aFunding` and `bFunding` to something more interesting that 1 and 0. Maybe 2 and 3? This would force `b` to make a deposit.

File: packages/server-wallet/src/wallet/wallet.ts
Comment (kerzhner) [inline packages/server-wallet/src/wallet/wallet.ts:227]: Curious, why is it important create a promise here instead of below, in the return dictionary declaration?

**Review by kerzhner — APPROVED** [review]
Thanks for putting this together!

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (lalexgap) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:188]: This should be waiting for the objective to complete fully.

In the result from both `approveObjectives` and `createChannels` we return a `done` promise that resolves when the objective is completed.

The idea is that you get a result right away with a channelId/ObjectiveId  and a `promise`  that will resolve when the objective is completed.

File: packages/server-wallet/src/wallet/wallet.ts
Comment (lalexgap) [reply packages/server-wallet/src/wallet/wallet.ts:227]: A few lines lower we call `await this.handleEngineOutput(closeResult)` which will emit events for objectives in `closeResult`.  If `this.handleEngineOutput(closeResult)` is called before we setup the promise then the objective event may be emitted **before** we construct the promise that resolves when the event is emitted.

That means we can end up with a promise that never resolves even though the objective is complete.

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (lalexgap) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts]: Yup, this is just some debugging code that snuck in.

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (lalexgap) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:198]: I did this in f42a5a9 which broke the tests! Thanks for bringing this up, investigating now.

Comment (lalexgap) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:199]: Ah it looks like it was due to the fact that we only `mine` blocks on the `Deposited` event. Since the `Deposited` event only fires when depositing funds, there's nothing to trigger mining a bunch of blocks when conclude is called.

   I'm guessing that with only one deposit the mining from the `Deposited` event was close enough to mine the conclude transaction as well so that's why we didn't see this with `aFunding=1, bFunding=0`.

This is fixed in https://github.com/statechannels/statechannels/pull/3557/commits/6f304b73de94b1ee13b763ae5d8095cad0e50218 by mining on the allocationUpdated event as well.

File: packages/server-wallet/jest/chain-setup.ts
Comment (lalexgap) [reply packages/server-wallet/jest/chain-setup.ts:13]: Nope! It attempts to create the directory and if it fails due to it existing (which would happen when running again) it swallows the error.

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (kerzhner) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:199]: If there was a single deposited event, in theory, the test should have failed. If I recall correctly, the test is set up to mine the minimum number of blocks for the test to pass.

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (kerzhner) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:184]: Got it! Thanks for the explanation!

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (kerzhner) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:193]: Resolved above.

File: packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts
Comment (lalexgap) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts:199]: > If there was a single deposited event, in theory, the test should have failed. If I recall correctly, the test is set up to mine the minimum number of blocks for the test to pass.

Ah you're right!

I think the reason it was passing was because `b` was the one calling `closeChannel`. Since it had `0` funding it would succeed as soon as the conclude proof was signed.

Comment (lalexgap) [reply packages/server-wallet/src/__chain-test__/directly-funded-channel.test.ts]: I've added some test cases in https://github.com/statechannels/statechannels/pull/3557/commits/40c47f5a0e3374fc5fbda10668010e4b66bc6e47 so we now test both A and B closing. I also threw in some cases with messages being dropped / delayed.

File: packages/server-wallet/src/wallet/wallet.ts
Comment (kerzhner) [reply packages/server-wallet/src/wallet/wallet.ts:227]: Thanks for the explanation.

**Pull request merged**
